rpm/packagescanner.go

```diff
@@ -161,78 +179,146 @@ func (ps *Scanner) Scan(ctx context.Context, layer *claircore.Layer) ([]*clairco
+	}
 
-			return s.Err()
-		}(); err != nil {
-			if errbuf.Len() != 0 {
-				zlog.Warn(ctx).
-					Str("db", db).
-					Strs("cmd", cmd.Args).
-					Str("err", errbuf.String()).
-					Msg("error output")
-			}
-			return nil, fmt.Errorf("rpm: error reading rpm output: %w", err)
+	l := &extractedLayer{root: root}
+	// This takes the extreme measure of panicing the program if the layer
+	// directory is leaked or unsuccessfully cleaned up.
+	//
+	// Since these can use a nontrivial amount of disk space (layers around 1GiB
+	// are not uncommon) the eventual result is the same, but this way operators
+	// are more likely to notice.
+	_, file, line, _ := runtime.Caller(1)
+	runtime.SetFinalizer(l, func(l *extractedLayer) {
+		panic(fmt.Sprintf("%s:%d: layer directory %q not closed", file, line, root))
+	})
+	extractedLayerProfile.Add(l, 2)
+	return l, nil
+}
+
+func (l *extractedLayer) Examine(ctx context.Context, pkgs *[]*claircore.Package, db string, kind int) error {
+	if err := ctx.Err(); err != nil {
+		return err
+	}
+	ctx = baggage.ContextWithValues(ctx,
+		label.String("component", "rpm/extractedLayer.Examine"),
+		label.String("db", db))
+	zlog.Debug(ctx).Msg("examining database")
+	var backend string
+	switch kind {
+	case backendBDB:
+		backend = `bdb_ro`
+	default:
+		panic(fmt.Sprintf("programmer error: called with bad kind: %v", kind))
+	}
+
+	cmd := exec.CommandContext(ctx, "rpm",
+		`--define`, `_db_backend `+backend,
+		`--root`, l.root, `--dbpath`, db,
+		`--query`, `--all`, `--queryformat`, queryFmt)
+	r, err := cmd.StdoutPipe()
+	if err != nil {
+		return err
+	}
+	// The Wait method closes this once the subprocess exits, so we only need to
+	// close it to signal we've gone away early.
+	errbuf := bytes.Buffer{}
+	cmd.Stderr = &errbuf
+	zlog.Debug(ctx).Strs("cmd", cmd.Args).Msg("rpm invocation")
+	if err := cmd.Start(); err != nil {
+		return err
+	}
+	defer func() {
+		if errbuf.Len() != 0 {
+			zlog.Warn(ctx).
+				Strs("cmd", cmd.Args).
+				Str("err", errbuf.String()).
+				Msg("error output")
 		}
-		if err := cmd.Wait(); err != nil {
-			return nil, err
+	}()
+
+	srcs := make(map[string]*claircore.Package)
+	s := bufio.NewScanner(r)
+	s.Split(querySplit)
+
+	for s.Scan() {
+		if err := ctx.Err(); err != nil {
+			return err
+		}
+		p, err := parsePackage(ctx, srcs, bytes.NewBuffer(s.Bytes()))
+		if err != nil {
+			if err := r.Close(); err != nil {
+				zlog.Warn(ctx).Err(err).Msg("error cleaning up subprocess")
+			}
+			return err
 		}
+		p.PackageDB = db
+		*pkgs = append(*pkgs, p)
 	}
 
-	return pkgs, nil
+	if err := s.Err(); err != nil {
+		return fmt.Errorf("rpm: error reading rpm output: %w", err)
+	}
+
+	if err := cmd.Wait(); err != nil {
+		return err
+	}
+	return nil
 }
 
+func (l *extractedLayer) Close() error {
+	if err := os.RemoveAll(l.root); err != nil {
+		return fmt.Errorf("rpm: will panic soon: unable to remove directory: %w", err)
+	}
+	runtime.SetFinalizer(l, nil)
```

Comment:
wow, super interesting way of raising a panic if `Close()` hasn't been called

Reply:
Yeah, I saw it here: https://crawshaw.io/blog/sharp-edged-finalizers

This use-case is a bit different, so maybe it's not appropriate, but between the additional flags telling tar to disregard permissions and ownership, setting permissions before doing work (instead of trying to do it just before cleanup), and the abstraction making the layer a bit easier to reason about, I think leaking this directory is either a programmer error or something seriously odd/missing in our handling.